daemon/emer-daemon.c

```diff
@@ -417,131 +497,95 @@ handle_http_response (SoupSession         *http_session,
 
-      GVariantBuilder event_id_builder;
-      get_uuid_builder (curr_sequence->event_id, &event_id_builder);
+  g_free (variants);
 
-      g_variant_builder_add (sequences_builder, "(uaya(xmv))",
-                             curr_sequence->user_id, &event_id_builder,
-                             &event_values_builder);
-      trash_sequence_event (curr_sequence);
-    }
-  priv->num_sequences_buffered = 0;
+  *read_bytes = curr_read_bytes;
 }
 
-/*
- * Populates the given GVariantBuilders with metric events from the persistent
- * cache. Returns TRUE on success even if there were no events found in the
- * persistent cache. Returns FALSE on failure. Regardless of success of failure,
- * the GVariantBuilders are guaranteed to be correctly initialized.
- */
-static gboolean
-drain_persistent_cache (EmerDaemonPrivate *priv,
-                        GVariantBuilder   *singulars_builder,
-                        GVariantBuilder   *aggregates_builder,
-                        GVariantBuilder   *sequences_builder)
+static void
+add_buffered_events_to_builders (EmerDaemon      *self,
+                                 gsize            num_bytes,
+                                 gsize           *num_variants,
+                                 GVariantBuilder *singulars,
+                                 GVariantBuilder *aggregates,
+                                 GVariantBuilder *sequences)
 {
-  g_variant_builder_init (singulars_builder, G_VARIANT_TYPE ("a(uayxmv)"));
-  g_variant_builder_init (aggregates_builder, G_VARIANT_TYPE ("a(uayxxmv)"));
-  g_variant_builder_init (sequences_builder, G_VARIANT_TYPE ("a(uaya(xmv))"));
-
-  GVariant **list_of_singulars;
-  GVariant **list_of_aggregates;
-  GVariant **list_of_sequences;
-
-  /*
-   * TODO: This value is currently unused by the persistent cache, but should be
-   * updated to a sensible value when the PC starts using it.
-   */
-  gint maximum_bytes_to_drain = 92160;
-  if (!emer_persistent_cache_drain_metrics (priv->persistent_cache,
-                                            &list_of_singulars,
-                                            &list_of_aggregates,
-                                            &list_of_sequences,
-                                            maximum_bytes_to_drain))
-    return FALSE;
-
-  for (gint i = 0; list_of_singulars[i] != NULL; i++)
-    g_variant_builder_add_value (singulars_builder, list_of_singulars[i]);
-  g_free (list_of_singulars);
-
-  for (gint i = 0; list_of_aggregates[i] != NULL; i++)
-    g_variant_builder_add_value (aggregates_builder, list_of_aggregates[i]);
-  g_free (list_of_aggregates);
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
 
-  for (gint i = 0; list_of_sequences[i] != NULL; i++)
-    g_variant_builder_add_value (sequences_builder, list_of_sequences[i]);
-  g_free (list_of_sequences);
+  gsize curr_num_variants = 0;
+  for (gsize curr_bytes = 0; curr_num_variants < priv->variant_array->len;
```

Comment:
I find it a little bit confusing to use a different variable in the loop than the one declared here...